File: cli/memory_tools.py
"""LLM tools для on-demand доступа к памяти Эйдос (P0–P1).

Включается при ``EIDOS_TOOLS=1`` и ``EIDOS_MEMORY_TOOLS`` не ``0``.
"""

from __future__ import annotations

import json
import os
import re
import time
from pathlib import Path
from typing import Any

from kernel.config import JOURNAL_DIR

_DEFAULT_EPISODIC_LIMIT = 20
_DEFAULT_SEMANTIC_LIMIT = 24
_DEFAULT_JOURNAL_TOP_K = 12
_DEFAULT_EXTERNAL_TOP_K = 8
_MAX_EPISODE_BODY = 48_000
_MAX_JOURNAL_FILE = 64_000


def memory_tools_enabled() -> bool:
    if os.environ.get("EIDOS_TOOLS", "1").strip().lower() in ("0", "false", "no", "off"):
        return False
    v = os.environ.get("EIDOS_MEMORY_TOOLS", "1").strip().lower()
    return v not in ("0", "false", "no", "off")


def memory_tool_specs() -> list[dict[str, Any]]:
    """OpenAI-compatible tool definitions."""
    return [
        {
            "type": "function",
            "function": {
                "name": "memory_search_episodic",
                "description": (
                    "Поиск в эпизодической памяти (SQLite episodes). "
                    "Укажите query (слова из реплики) и/или cues (список ключей). "
                    "Без query — последние эпизоды по времени."
                ),
                "parameters": {
                    "type": "object",
                    "properties": {
                        "query": {"type": "string", "description": "Подстрока в summary/raw_text"},
                        "cues": {
                            "type": "array",
                            "items": {"type": "string"},
                            "description": "Ключи context_keys (как в recall_by_cues)",
                        },
                        "limit": {"type": "integer", "description": "Макс. записей (1–80)"},
                        "min_salience": {"type": "number"},
                    },
                },
            },
        },
        {
            "type": "function",
            "function": {
                "name": "memory_get_episode",
                "description": "Получить один эпизод episodic по числовому id (полный raw/summary).",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "episode_id": {"type": "integer", "description": "id из episodic"},
                    },
                    "required": ["episode_id"],
                },
            },
        },
        {
            "type": "function",
            "function": {
                "name": "memory_list_semantic",
                "description": "Список семантических принципов (knowledge.db), с опциональным фильтром.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "query": {"type": "string", "description": "Подстрока в тексте принципа"},
                        "limit": {"type": "integer"},
                        "min_confidence": {"type": "number"},
                    },
                },
            },
        },
        {
            "type": "function",
            "function": {
                "name": "memory_search_journal",
                "description": "Поиск по markdown-дневнику (data/journal/*.md).",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "query": {"type": "string"},
                        "top_k": {"type": "integer"},
                        "mode": {
                            "type": "string",
                            "enum": ["auto", "semantic", "lexical"],
                            "description": "auto: rubert если индекс есть, иначе substring",
                        },
                    },
                    "required": ["query"],
                },
            },
        },
        {
            "type": "function",
            "function": {
                "name": "memory_read_journal",
                "description": "Прочитать файл дневника (только имя, напр. 2025-05-20.md).",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "filename": {"type": "string"},
                        "max_chars": {"type": "integer"},
                    },
                    "required": ["filename"],
                },
            },
        },
        {
            "type": "function",
            "function": {
                "name": "memory_search_external",
                "description": (
                    "Векторный поиск по внешней памяти (documents.db + index). "
                    "Требует проиндексированный корпус."
                ),
                "parameters": {
                    "type": "object",
                    "properties": {
                        "query": {"type": "string"},
                        "top_k": {"type": "integer"},
                        "min_score": {"type": "number"},
                    },
                    "required": ["query"],
                },
            },
        },
    ]


def is_memory_tool_name(name: str) -> bool:
    return name.startswith("memory_")


def _cue_tokens(text: str | None) -> list[str]:
    if not text or not str(text).strip():
        return []
    words = re.findall(r"[\wА-Яа-яЁё-]{3,}", str(text).lower())
    out: list[str] = []
    seen: set[str] = set()
    for w in words:
        if w in seen:
            continue
        seen.add(w)
        out.append(w)
        if len(out) >= 36:
            break
    return out


def _episode_by_id(episodic: Any, episode_id: int) -> dict[str, Any] | None:
    rows = episodic._conn.execute(  # noqa: SLF001
        "SELECT * FROM episodes WHERE id = ?",
        (episode_id,),
    ).fetchall()
    if not rows:
        return None
    columns = [
        d[1] for d in episodic._conn.execute("PRAGMA table_info(episodes)").fetchall()  # noqa: SLF001
    ]
    return dict(zip(columns, rows[0]))


def _truncate_episode(ep: dict[str, Any]) -> dict[str, Any]:
    out = dict(ep)
    for key in ("raw_text", "summary", "compressed"):
        val = str(out.get(key) or "")
        if len(val) > 1200:
            out[key] = val[:1199] + "…"
    return out


def execute_memory_tool(memory: Any, name: str, args: dict[str, Any]) -> str:
    """Выполнить memory_*; возвращает JSON-строку для роли tool."""
    try:
        if name == "memory_search_episodic":
            payload = _search_episodic(memory, args)
        elif name == "memory_get_episode":
            payload = _get_episode(memory, args)
        elif name == "memory_list_semantic":
            payload = _list_semantic(memory, args)
        elif name == "memory_search_journal":
            payload = _search_journal(memory, args)
        elif name == "memory_read_journal":
            payload = _read_journal(args)
        elif name == "memory_search_external":
            payload = _search_external(memory, args)
        else:
            payload = {"error": f"неизвестный memory tool: {name}"}
        return json.dumps(payload, ensure_ascii=False, indent=0)
    except Exception as exc:
        return json.dumps({"error": str(exc)}, ensure_ascii=False)


def _search_episodic(memory: Any, args: dict[str, Any]) -> dict[str, Any]:
    limit = int(args.get("limit") or _DEFAULT_EPISODIC_LIMIT)
    limit = max(1, min(limit, 80))
    min_salience = float(args.get("min_salience") or 0.0)
    query = str(args.get("query") or "").strip()
    cues_raw = args.get("cues")
    cues: list[str] = []
    if isinstance(cues_raw, list):
        cues = [str(c).strip().lower() for c in cues_raw if str(c).strip()]
    elif query:
        cues = _cue_tokens(query)

    episodic = memory.episodic
    hits: list[dict[str, Any]] = []

    if cues:
        hits = episodic.recall_by_cues(cues, limit=limit)
    elif query:
        q = query.lower()
        pool = episodic.query_all(min_salience=min_salience, max_rows=5000)
        scored: list[tuple[float, dict[str, Any]]] = []
        for ep in pool:
            body = (
                str(ep.get("raw_text") or "")
                + " "
                + str(ep.get("summary") or "")
            ).lower()
            if q in body:
                scored.append((float(ep.get("timestamp") or 0), ep))
        scored.sort(reverse=True)
        hits = [ep for _, ep in scored[:limit]]
    else:
        hits = episodic.query(limit=limit, min_salience=min_salience)

    return {
        "count": len(hits),
        "episodes": [_truncate_episode(ep) for ep in hits],
    }


def _get_episode(memory: Any, args: dict[str, Any]) -> dict[str, Any]:
    eid = int(args.get("episode_id") or 0)
    if eid <= 0:
        return {"error": "episode_id required"}
    ep = _episode_by_id(memory.episodic, eid)
    if not ep:
        return {"error": f"episode {eid} not found"}
    for key in ("raw_text", "summary"):
        val = str(ep.get(key) or "")
        if len(val) > _MAX_EPISODE_BODY:
            ep[key] = val[: _MAX_EPISODE_BODY - 1] + "…"
    return {"episode": ep}


def _list_semantic(memory: Any, args: dict[str, Any]) -> dict[str, Any]:
    limit = int(args.get("limit") or _DEFAULT_SEMANTIC_LIMIT)
    limit = max(1, min(limit, 100))
    min_conf = float(args.get("min_confidence") or 0.0)
    query = str(args.get("query") or "").strip().lower()
    rows = memory.semantic.get_principles(min_confidence=min_conf)
    if query:
        rows = [
            r
            for r in rows
            if query in str(r.get("principle") or "").lower()
        ]
    rows = rows[:limit]
    return {"count": len(rows), "principles": rows}


def _search_journal(_memory: Any, args: dict[str, Any]) -> dict[str, Any]:
    from kernel.journal import Journal

    query = str(args.get("query") or "").strip()
    if not query:
        return {"error": "query required"}
    top_k = int(args.get("top_k") or _DEFAULT_JOURNAL_TOP_K)
    top_k = max(1, min(top_k, 30))
    mode = str(args.get("mode") or "auto").strip().lower()
    j = Journal(memory=None)
    if mode == "lexical":
        hits = j.search_substring(query)
    elif mode == "semantic":
        hits = j.search_semantic(query, top_k=top_k, fallback=True)
    else:
        hits = j.search_semantic(query, top_k=top_k, fallback=True)
    return {"count": len(hits), "hits": hits[:top_k]}


def _read_journal(args: dict[str, Any]) -> dict[str, Any]:
    name = str(args.get("filename") or "").strip().replace("\\", "/")
    if not name or ".." in name or "/" in name.strip("/"):
        return {"error": "filename must be a bare .md name, e.g. 2025-05-20.md"}
    if not name.endswith(".md"):
        name = f"{name}.md"
    path = (JOURNAL_DIR / name).resolve()
    if not path.is_file() or path.parent != JOURNAL_DIR.resolve():
        return {"error": f"not found: {name}"}
    text = path.read_text(encoding="utf-8", errors="replace")
    max_chars = int(args.get("max_chars") or _MAX_JOURNAL_FILE)
    max_chars = max(1000, min(max_chars, 200_000))
    truncated = len(text) > max_chars
    if truncated:
        text = text[: max_chars - 1] + "…"
    return {"filename": name, "chars": len(text), "truncated": truncated, "content": text}


def _search_external(memory: Any, args: dict[str, Any]) -> dict[str, Any]:
    query = str(args.get("query") or "").strip()
    if not query:
        return {"error": "query required"}
    top_k = int(args.get("top_k") or _DEFAULT_EXTERNAL_TOP_K)
    top_k = max(1, min(top_k, 20))
    min_score = float(args.get("min_score") or 0.12)
    try:
        hits = memory.external.search(query, top_k=top_k, min_score=min_score)
    except Exception as exc:
        return {"error": f"external search failed: {exc}", "hits": []}
    return {"count": len(hits), "hits": hits}


def format_memory_help() -> str:
    """Текст для slash ``/memory``."""
    pipe = os.environ.get("EIDOS_CHAT_MEMORY_PIPELINE", "full").strip() or "full"
    active = os.environ.get("EIDOS_CHAT_ACTIVE_MEMORY", "1").strip().lower() not in (
        "0",
        "false",
        "no",
        "off",
    )
    tools_on = memory_tools_enabled()
    lines = [
        "Память Эйдос — два контура",
        "",
        "1) Пассивно (каждый ход): блок «Активное извлечение» в system.",
        f"   pipeline={pipe}, active_memory={'on' if active else 'off'}",
        "   Env: EIDOS_CHAT_MEMORY_PIPELINE, EIDOS_CHAT_ACTIVE_MEMORY_CHARS, …",
        "",
        "2) По запросу (tools):",
        f"   memory tools={'on' if tools_on else 'off'} (EIDOS_MEMORY_TOOLS)",
    ]
    if tools_on:
        for spec in memory_tool_specs():
            fn = spec.get("function") or {}
            lines.append(f"   · {fn.get('name', '?')}")
    else:
        lines.append("   (выключено: EIDOS_MEMORY_TOOLS=0)")
    lines.append("")
    lines.append("Документация: docs/MEMORY_AGENT_ACCESS.md")
    return "\n".join(lines)

File: logos-rs/crates/eidos-core/src/memory_tools.rs
//! On-demand memory tools (P2): Rust fallback when Python sidecar unavailable.

use std::fs;
use serde_json::{json, Value};

use crate::episodic_store::EpisodicStore;
use crate::error::{CoreError, Result};
use crate::paths::Paths;
use crate::semantic_store::SemanticStore;

const MAX_EPISODE_BODY: usize = 48_000;
const MAX_JOURNAL_FILE: usize = 64_000;

pub fn memory_tools_enabled() -> bool {
    let v = std::env::var("EIDOS_TOOLS")
        .unwrap_or_else(|_| "1".into())
        .to_ascii_lowercase();
    if matches!(v.as_str(), "0" | "false" | "no" | "off") {
        return false;
    }
    let m = std::env::var("EIDOS_MEMORY_TOOLS")
        .unwrap_or_else(|_| "1".into())
        .to_ascii_lowercase();
    !matches!(m.as_str(), "0" | "false" | "no" | "off")
}

pub fn is_memory_tool_name(name: &str) -> bool {
    name.starts_with("memory_")
}

pub fn memory_tool_specs() -> Vec<Value> {
    vec![
        json!({
            "type": "function",
            "function": {
                "name": "memory_search_episodic",
                "description": "Поиск в episodic SQLite (Rust fallback; без external vectors).",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "query": { "type": "string" },
                        "limit": { "type": "integer" }
                    }
                }
            }
        }),
        json!({
            "type": "function",
            "function": {
                "name": "memory_get_episode",
                "description": "Эпизод по id.",
                "parameters": {
                    "type": "object",
                    "properties": { "episode_id": { "type": "integer" } },
                    "required": ["episode_id"]
                }
            }
        }),
        json!({
            "type": "function",
            "function": {
                "name": "memory_list_semantic",
                "description": "Список принципов semantic DB.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "query": { "type": "string" },
                        "limit": { "type": "integer" }
                    }
                }
            }
        }),
        json!({
            "type": "function",
            "function": {
                "name": "memory_search_journal",
                "description": "Лексический поиск по journal/*.md (Rust).",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "query": { "type": "string" },
                        "top_k": { "type": "integer" }
                    },
                    "required": ["query"]
                }
            }
        }),
        json!({
            "type": "function",
            "function": {
                "name": "memory_read_journal",
                "description": "Прочитать файл дневника.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "filename": { "type": "string" },
                        "max_chars": { "type": "integer" }
                    },
                    "required": ["filename"]
                }
            }
        }),
    ]
}

pub fn execute_memory_tool(paths: &Paths, name: &str, arguments_json: &str) -> String {
    match execute_memory_tool_inner(paths, name, arguments_json) {
        Ok(v) => serde_json::to_string(&v).unwrap_or_else(|e| json!({ "error": e.to_string() }).to_string()),
        Err(e) => json!({ "error": e.to_string() }).to_string(),
    }
}

fn execute_memory_tool_inner(paths: &Paths, name: &str, arguments_json: &str) -> Result<Value> {
    let args: Value = serde_json::from_str(arguments_json).unwrap_or(json!({}));
    match name {
        "memory_search_episodic" => search_episodic(paths, &args),
        "memory_get_episode" => get_episode(paths, &args),
        "memory_list_semantic" => list_semantic(paths, &args),
        "memory_search_journal" => search_journal(paths, &args),
        "memory_read_journal" => read_journal(paths, &args),
        "memory_search_external" => Ok(json!({
            "error": "external search requires Python sidecar (vectors)",
            "hits": []
        })),
        _ => Err(CoreError::Tool(format!("неизвестный memory tool: {name}"))),
    }
}

fn search_episodic(paths: &Paths, args: &Value) -> Result<Value> {
    let store = EpisodicStore::open(paths)?;
    let limit = args
        .get("limit")
        .and_then(|v| v.as_u64())
        .unwrap_or(20)
        .clamp(1, 80) as usize;
    let query = args
        .get("query")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim()
        .to_lowercase();
    let mut hits = if query.is_empty() {
        store.query(limit, 0.0)?
    } else {
        let pool = store.query(5000, 0.0)?;
        pool.into_iter()
            .filter(|ep| {
                let body = format!("{} {}", ep.summary, ep.raw_text).to_lowercase();
                body.contains(&query)
            })
            .take(limit)
            .collect()
    };
    let episodes: Vec<Value> = hits
        .drain(..)
        .map(|ep| {
            json!({
                "id": ep.id,
                "timestamp": ep.timestamp,
                "salience": ep.salience,
                "summary": truncate_str(&ep.summary, 1200),
                "raw_text": truncate_str(&ep.raw_text, 1200),
            })
        })
        .collect();
    Ok(json!({ "count": episodes.len(), "episodes": episodes, "backend": "rust" }))
}

fn get_episode(paths: &Paths, args: &Value) -> Result<Value> {
    let id = args
        .get("episode_id")
        .and_then(|v| v.as_i64())
        .ok_or_else(|| CoreError::Tool("episode_id required".into()))?;
    let store = EpisodicStore::open(paths)?;
    let ep = store
        .get_by_id(id)?
        .ok_or_else(|| CoreError::Tool(format!("episode {id} not found")))?;
    Ok(json!({
        "episode": {
            "id": ep.id,
            "timestamp": ep.timestamp,
            "summary": truncate_str(&ep.summary, MAX_EPISODE_BODY),
            "raw_text": truncate_str(&ep.raw_text, MAX_EPISODE_BODY),
        },
        "backend": "rust"
    }))
}

fn list_semantic(paths: &Paths, args: &Value) -> Result<Value> {
    let store = SemanticStore::open(paths)?;
    let limit = args
        .get("limit")
        .and_then(|v| v.as_u64())
        .unwrap_or(24)
        .clamp(1, 100) as usize;
    let query = args
        .get("query")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim()
        .to_lowercase();
    let mut rows = store.get_principles(0.0, 500)?;
    if !query.is_empty() {
        rows.retain(|r| r.principle.to_lowercase().contains(&query));
    }
    rows.truncate(limit);
    let principles: Vec<Value> = rows
        .iter()
        .map(|r| {
            json!({
                "id": r.id,
                "principle": truncate_str(&r.principle, 2000),
                "confidence": r.confidence,
            })
        })
        .collect();
    Ok(json!({ "count": principles.len(), "principles": principles, "backend": "rust" }))
}

fn search_journal(paths: &Paths, args: &Value) -> Result<Value> {
    let query = args
        .get("query")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim()
        .to_lowercase();
    if query.is_empty() {
        return Err(CoreError::Tool("query required".into()));
    }
    let top_k = args
        .get("top_k")
        .and_then(|v| v.as_u64())
        .unwrap_or(12)
        .clamp(1, 30) as usize;
    let dir = paths.journal_dir();
    let mut hits: Vec<Value> = Vec::new();
    if dir.is_dir() {
        for ent in fs::read_dir(&dir).map_err(|e| CoreError::WorkingMemory(e.to_string()))? {
            let ent = ent.map_err(|e| CoreError::WorkingMemory(e.to_string()))?;
            let path = ent.path();
            if path.extension().and_then(|e| e.to_str()) != Some("md") {
                continue;
            }
            if path.file_name().and_then(|n| n.to_str()) == Some("INDEX.md") {
                continue;
            }
            let text = fs::read_to_string(&path).unwrap_or_default().to_lowercase();
            if text.contains(&query) {
                let name = path.file_name().and_then(|n| n.to_str()).unwrap_or("?");
                hits.push(json!({ "file": name, "snippet": query }));
            }
            if hits.len() >= top_k {
                break;
            }
        }
    }
    Ok(json!({ "count": hits.len(), "hits": hits, "backend": "rust_lexical" }))
}

fn read_journal(paths: &Paths, args: &Value) -> Result<Value> {
    let mut name = args
        .get("filename")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim()
        .replace('\\', "/");
    if name.is_empty() || name.contains("..") {
        return Err(CoreError::Tool("filename required".into()));
    }
    if !name.ends_with(".md") {
        name.push_str(".md");
    }
    let path = paths.journal_dir().join(&name);
    let canon = path.canonicalize().map_err(|_| CoreError::NotFound(path.clone()))?;
    let dir = paths
        .journal_dir()
        .canonicalize()
        .unwrap_or_else(|_| paths.journal_dir());
    if !canon.starts_with(&dir) {
        return Err(CoreError::Tool("path outside journal".into()));
    }
    let text = fs::read_to_string(&canon).map_err(|_| CoreError::NotFound(path))?;
    let max_chars = args
        .get("max_chars")
        .and_then(|v| v.as_u64())
        .unwrap_or(MAX_JOURNAL_FILE as u64)
        .clamp(1000, 200_000) as usize;
    let truncated = text.len() > max_chars;
    let content = if truncated {
        format!("{}…", &text[..max_chars.saturating_sub(1)])
    } else {
        text
    };
    Ok(json!({
        "filename": name,
        "chars": content.len(),
        "truncated": truncated,
        "content": content,
        "backend": "rust"
    }))
}

fn truncate_str(s: &str, max: usize) -> String {
    if s.len() <= max {
        s.to_string()
    } else {
        format!("{}…", &s[..max.saturating_sub(1)])
    }
}

/// Краткая справка для ``/memory`` (без Python).
pub fn format_memory_help_rust() -> String {
    let tools = memory_tools_enabled();
    let mut lines = vec![
        "Память Эйдос (справка)".to_string(),
        String::new(),
        "Пассивно: блок active_memory в system (Python sidecar при обычном чате).".to_string(),
        format!("Tools memory_*: {}", if tools { "вкл." } else { "выкл." }),
        String::new(),
        "Документация: docs/MEMORY_AGENT_ACCESS.md".to_string(),
    ];
    if tools {
        for spec in memory_tool_specs() {
            if let Some(name) = spec
                .get("function")
                .and_then(|f| f.get("name"))
                .and_then(|v| v.as_str())
            {
                lines.push(format!("  · {name}"));
            }
        }
    }
    lines.join("\n")
}
